mitodl/mit-open
# Add RestrictedRoute for controling route access

# What are the relevant tickets?
Closes https://github.com/mitodl/mit-open/issues/228

# Description (What does it do?)
This PR adds a `<RestrictedRoute allow={user => boolean} />` component that can be used to restrict access to frontend routes. It behaves the same as 403 API errors: shows authenticated users the ForbiddenPage, and redirects unauthenticated users to login.

# How can this be tested?
**Prerequesite:** You should have access to a user that is NOT an admin/superuser. You can create one via the ui at http://localhost:8063/login/ol-oidc/ if you have keycloak creds from RC (I can share). Alternatively you can create a superuser then degrade its permissions via django shell:
```
./manage.py createsuperuser
# finish superuser creation
./manage.py shell
# In python
from django.contrib.auth import get_user_model
User = get_user_model()
user = User.objects.get(username="whatever_username")
user.is_admin = False
user.is_superuser = False 
user.save()
```

1. As an unauthenticated user, visit http://localhost:8063/articles/new ; you should be redirected to login
2. As an authenticated user WITHOUT permission (non-superuser, non-admin), visit http://localhost:8063/articles/new 
    - in dev mode, you'll see a black overlay from webpack. Click the X in upper right corner
    - Now you should see the ForbiddenPage.
3. As an authenticated user with permission (e.g., superuser), visit http://localhost:8063/articles/new. You should see the article creation page.
